pysynchronizer/icbinn.py

```diff
@@ -0,0 +1,241 @@
+#
+# Copyright (c) 2013 Citrix Systems, Inc.
+# Copyright (c) 2021 Daniel P. Smith, Apertus Solutions LLC
+# 
+# This program is free software; you can redistribute it and/or modify
+# it under the terms of the GNU General Public License as published by
+# the Free Software Foundation; either version 2 of the License, or
+# (at your option) any later version.
+# 
+# This program is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU General Public License for more details.
+# 
+# You should have received a copy of the GNU General Public License
+# along with this program; if not, write to the Free Software
+# Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA
+#
+
+
+from os.path import join
+from pathlib import Path
+
+from .errors import InsufficientIcbinnPaths, IcbinnConnectError, PlatformError, IcbinnError, TargetStateError
+from .xenstore import Xenstore
+from .oxt_dbus import OXTDBusApi
+from .singleton import Singleton
+
+from pyicbinn import icbinn_clnt_create_argo, icbinn_close, icbinn_lock
+from pyicbinn import icbinn_mkdir, icbinn_open, icbinn_pwrite, icbinn_rand
+from pyicbinn import icbinn_readent, icbinn_rename, icbinn_stat, icbinn_pread
+from pyicbinn import icbinn_unlink
+
+ICBINN_SERVER_PORT = 4878
+ICBINN_SERVER_DOMAIN_ID = 0
+ICBINN_MAXDATA = 65536
+ICBINN_FILE = 0
+ICBINN_DIRECTORY = 1
+ICBINN_UNKNOWN = 2
+ICBINN_LTYPE_RDLCK = 0
+ICBINN_LTYPE_WRLCK = 1
+ICBINN_LTYPE_UNLCK = 2
+ICBINN_RANDOM = 0
+ICBINN_URANDOM = 1
+
+DISK_DIR = 'disks'
+ENCRYPT_SNAPSHOTS = 'encrypt_snapshots'
+STATUS_OKAY = 0
+STATUS_INTERNAL_EXCEPTION = 1
+STATUS_FAILED = 2
+DOWNLOAD_BLOCK_SIZE = 512 * 1024
+ENCRYPTION_KEY_BYTES = 64
+PROGRESS_INTERVAL = 1
+DISK_TYPE_ISO = 'iso'
+DISK_TYPE_VHD = 'vhd'
+DISK_TYPES = [DISK_TYPE_ISO, DISK_TYPE_VHD]
+SYNC_ROLE_PLATFORM = 'platform'
+SYNC_ROLE_REALM = 'realm'
+SYNC_ROLES = [SYNC_ROLE_PLATFORM, SYNC_ROLE_REALM]
+RPC_PREFIX = 'rpc:'
+
+class Icbinn(Singleton):
+    def __init__(self):
+        uuid_node = Path('/sys/hypervisor/uuid')
+        if uuid_node.exists():
+            output = "/vm/" + uuid_node.read_text()
+        else:
+            output = Xenstore.read('vm').rstrip()
+
+        syncvm_object_path = output.replace('-', '_')
+        paths = OXTDBusApi.open_vm(syncvm_object_path).get_property('icbinn-path').split(',')
+
+        if len(paths) != 2:
+            raise InsufficientIcbinnPaths(2, paths)
+
+        self.storage = IcbinnRemote(paths[0], server_port=ICBINN_SERVER_PORT + 0)
+        self.config = IcbinnRemote(paths[1], server_port=ICBINN_SERVER_PORT + 1)
```

Comment:
I realize you copied it over from sync-client, but self.config doesn't seem to be used.

Reply:
Yep, and as far as I can recall, I don't remember seeing it used in sync-client either. I can comment out for now as it will drop an unnecessary Argo connection.